Add Data - CSV Upload UI

Bargs opened the pull request:
Fixes #6541 
Depends on #6844 and #6842 

This PR implements the CSV Upload Add Data Wizard. Some of the highlights of the changes:
- A new `upload-wizard` directive which acts as the container for the CSV wizard. This intentionally duplicates some code and structure from the Tail a File wizard to avoid prematurely DRYing things up before we really know exactly what a wizard API needs to look like.
- A new `parse-csv-step` directive which allows a user to pick and preview their CSV
- A new `upload-data-step` directive for uploading the CSV to the [Kibana _data API](https://github.com/elastic/kibana/pull/6844) and displaying the results to the user
- A new `uploadCSV` method on the ingest service, for sending File objects to the Kibana _data API.

The CSV wizard does **not** currently include a pipeline creation step. It will be re-added once we have editable pipelines implemented.

rashidkpc commented:
jenkins, test it

rashidkpc commented:
@tsullivan can you look at the test failure here? It seems to be related to the uuid code

tsullivan commented:
Looking into it. I get the failure locally too

tsullivan commented:
@rashidkpc looks like all the tests that use this mechanism cause a failure:
```
    kbnServer = kbnTestServer.createServer({
      plugins: {
        scanDirs: [
          fromRoot('src/plugins')
        ]
      }
    });
```

The server test suite only passes if I skip all these:
- `src/plugins/elasticsearch/lib/__tests__/manage_uuid.js`
- `src/plugins/elasticsearch/lib/__tests__/routes.js`
- `src/server/http/__tests__/index.js`

My theory is that the scanDirs/fromRoot setup is calls some constructor function without the `new` keyword, which causes context to be bound to the global context. But I'm still not really sure. Still looking.

tsullivan commented:
@Bargs looks like the highland module is purposefully creating a global variable called `nil` and that's why mocha is failing with `Error: global leak detected: nil`

Source: https://github.com/caolan/highland/blob/master/dist/highland.js#L177

Looks like there is a way to register globals via config with grunt-simple-mocha: https://github.com/yaymukund/grunt-simple-mocha

Or you can add a line with `--globals nil` to `test/mocha.opts`

tsullivan commented:
Has conflicts that need to be resolved

Bargs commented:
Thanks @tsullivan, adding nil to the list of globals fixed the issue.

Bargs commented:
jenkins, test it

Bargs commented:
@BigFunger I think this one is ready for code review. @alt74 may still have a few design tweaks it seems like they'll be pretty small.

BigFunger commented:
- 'Tail a File' is still an option. I assume it just needs to be merged with `feature/ingest`
- A few issues related to the **Index name or pattern** field
 - Still defaults to `filebeat-*`, can we suggest an index name that is based on the filename they are trying to upload?
 - Data import process fails if the value in this field has invalid characters for an index. (which is a value that the field defaults to)
 - Data import process failes if there are capitol letters
 - Basically, needs some field validation.
- The review page fails if there are more than one error messages that are identical. Add an `track by $index` clause to the ng-repeat value. I was able to see this issue when I didn't change the default index name (with an asterix), which caused an error per line in the source csv
- When there was one error per line in the input file, this caused many errors, which makes the page scroll a lot. Do we want unlimited scrolling here, or paged data?
- Before I figured out that the page was failing because of identical error values, the only message I was able to see is 'We encountered errors while indexing your data.' I assume this will be resolved when adding the `track by' clause.
- Clicking the 'Done' button on the last step brings me to the discover app, but...
 - the index that I just created does not show up until I refresh the page.
 - the index that I just created is not selected.

BigFunger commented on src/plugins/kibana/public/settings/sections/indices/upload/directives/upload_wizard.js:
```diff
@@ -0,0 +1,95 @@
+import IngestProvider from 'ui/ingest';
+import 'plugins/kibana/settings/sections/indices/add_data_steps/pattern_review_step';
+import 'plugins/kibana/settings/sections/indices/add_data_steps/parse_csv_step';
+import 'plugins/kibana/settings/sections/indices/add_data_steps/pipeline_setup';
```
sadly, this should be removed for now.

BigFunger commented on src/plugins/kibana/public/settings/sections/indices/upload/directives/upload_wizard.js:
```diff
@@ -0,0 +1,95 @@
+      scope: {},
+      bindToController: true,
+      controllerAs: 'wizard',
+      controller: function ($scope, AppState, safeConfirm, kbnUrl, $http, Notifier, $window, config, Private) {
```
Doesn't look like `$http` or `config` dependencies are used. (also true in `filebeat_wizard`)

Bargs commented:
> The review page fails if there are more than one error messages that are identical. Add an track by $index clause to the ng-repeat value. I was able to see this issue when I didn't change the default index name (with an asterix), which caused an error per line in the source csv

Do you mean the final page where the csv is uploaded and the user gets the results?

> When there was one error per line in the input file, this caused many errors, which makes the page scroll a lot. Do we want unlimited scrolling here, or paged data?

Getting an error for every line should hopefully be a rare occurrence. Maybe we should just cap the list of line by line error messages at 5-10 and have a "See more" link to expand the alert box? I don't think I'd want paginated error messages.

> Clicking the 'Done' button on the last step brings me to the discover app, but...

I spent some time making sure the index pattern list gets updated without refresh, so I'll have to look into what's happening there.

Do you know if there's a way to generate a link to an app with certain app state pre-set? That's what I need to do to set the selected index pattern when linking to Discover from the Done button. I posed the question to the team before but didn't get an answer.

BigFunger commented:
> Do you mean the final page where the csv is uploaded and the user gets the results?

No. I mean the `Upload` step where the errors would be shown.

> Getting an error for every line should hopefully be a rare occurrence. Maybe we should just cap the list of line by line error messages at 5-10 and have a "See more" link to expand the alert box? I don't think I'd want paginated error messages.

I agree, this would work. @alt74?

> Do you know if there's a way to generate a link to an app with certain app state pre-set? That's what I need to do to set the selected index pattern when linking to Discover from the Done button. I posed the question to the team before but didn't get an answer.

There may be a better way, but you should be able to just set it in the URL: ie: ``kbnUrl.change(`/discover?_a=(index:${this.stepResults.indexPattern.id})`);``
(note: that this *doesn't* work right now, because of the before mentioned issue with the index not showing up until hitting refresh)

BigFunger commented:
>This intentionally duplicates some code and structure from the Tail a File wizard to avoid prematurely DRYing things up before we really know exactly what a wizard API needs to look like.

I created #7216 to address this. I didn't see an issue for it.

BigFunger commented on src/plugins/kibana/public/settings/sections/indices/add_data_steps/parse_csv_step/parse_csv_step.js:
```diff
@@ -0,0 +1,79 @@
+                    return `${error.type} at row ${error.row} - ${error.message}`;
+                  });
+                  this.columns = results.meta.fields;
+                  this.rows = _.map(results.data, _.values);
```
Is this a potential bug? 
From the lodash documentation
> ```_.values(new Foo);
// → [1, 2] (iteration order is not guaranteed)```

Wouldn't this imply that the order of the fields that are being mapped could be in a different order from row to row in the results?

Bargs replied:
Yup, great catch

Bargs commented:
> No. I mean the Upload step where the errors would be shown.

Yeah, I believe we're talking about the same page. I was just confused because you originally called it the "review page".

> There may be a better way, but you should be able to just set it in the URL: ie: kbnUrl.change(`/discover?_a=(index:${this.stepResults.indexPattern.id})`);

I definitely don't want to hard code the app state, it would break if the app state encoding scheme ever changed. I'll ask around some more and see what I can do.

BigFunger commented on src/plugins/kibana/public/settings/sections/indices/add_data_steps/parse_csv_step/parse_csv_step.js:
```diff
@@ -0,0 +1,79 @@
+            this.parseOptions
+          );
+
+          Papa.parse(this.file, config);
```
Should this be wrapped in a try/catch? I don't know enough about Papaparse, but I assume this would be a good place to wrap.

Bargs replied:
As far as I know papaparse doesn't throw any errors, it returns them as a part of the results.

BigFunger commented on src/plugins/kibana/public/settings/sections/indices/add_data_steps/upload_data_step/upload_data_step.js:
```diff
@@ -0,0 +1,44 @@
+import IngestProvider from 'ui/ingest';
+
+modules.get('apps/settings')
+  .directive('uploadDataStep', function () {
```
This block shouldn't be indented.

BigFunger commented on src/plugins/kibana/public/settings/sections/indices/add_data_steps/upload_data_step/upload_data_step.js:
```diff
@@ -0,0 +1,44 @@
+      },
+      bindToController: true,
+      controllerAs: 'uploadStep',
+      controller: function ($scope, $http, Notifier, $window, Private) {
```
`$scope`, `$http` are not used.

BigFunger commented on src/plugins/kibana/public/settings/sections/indices/add_data_steps/upload_data_step/upload_data_step.js:
```diff
@@ -0,0 +1,44 @@
+          location: 'Add Data'
+        });
+
+        const usePipeline = !_.isEmpty(_.get(this.results, 'pipeline.processors'));
```
Do we want to leave this pipeline code in here?

Bargs replied:
Yeah I'd like to leave it, it'll correctly return false for now and I'd only have to rewrite it when pipelines are re-added.

BigFunger commented on src/plugins/kibana/public/settings/sections/indices/upload/directives/upload_wizard.js:
```diff
@@ -0,0 +1,95 @@
+import 'plugins/kibana/settings/sections/indices/add_data_steps/parse_csv_step';
+import 'plugins/kibana/settings/sections/indices/add_data_steps/pipeline_setup';
+import 'plugins/kibana/settings/sections/indices/add_data_steps/upload_data_step';
+
```
Shouldn't this still import `import '../../styles/_add_data_wizard.less';` since the markup depends on it?

BigFunger commented on src/plugins/kibana/public/settings/sections/indices/add_data_steps/parse_csv_step/parse_csv_step.html:
```diff
@@ -0,0 +1,57 @@
+      </thead>
+      <tbody>
+      <tr ng-repeat="row in wizard.rows">
+        <td ng-repeat="cell in row track by $index">{{ cell }}</td>
```
Why are we limiting the length of the column headers to 12 characters, but not limiting the data (which in theory could be significantly longer than 12 characters)?

Bargs replied:
Because the data can wrap but the columns shouldn't. At least that's my thinking. The column limit idea came from @alt74, what do you think Jurgen?

BigFunger replied:
It *can* wrap, but the css as it is now will not force it to wrap if it gets big.

Bargs replied:
You're right, right now it's left up to the browser to figure out the best sizing for each column. Is there a better way to handle it?

BigFunger commented on src/plugins/kibana/public/settings/sections/indices/add_data_steps/parse_csv_step/parse_csv_step.html:
```diff
@@ -0,0 +1,57 @@
+    <button class="btn btn-primary btn-lg controls upload" ng-click>
+      Select File
+    </button>
+    <div>Maximum upload file size: 1 GB</div>
```
Is this value available from config, etc so that it doesn't get out of sync with the actual limit?

Bargs replied:
Unfortunately it's only available server side.

BigFunger commented on src/plugins/kibana/public/settings/sections/indices/upload/directives/upload_wizard.html:
```diff
@@ -0,0 +1,86 @@
+<div class="wizard-container">
```
There seem to be some tab/spacing issues in this markup.

Bargs commented:
@BigFunger I'm still working through some of your feedback but I just pushed a good chunk of changes you might want to start taking a look at. I addressed most of the issues you mentioned in your first comment:

* Tail a file is no longer an option
* Index name now gets proper validation
* Default index name is now `data` instead of `filebeat-*`. I tried making it the filename, but realized many filenames are going to be invalid index patterns (special characters, whitespace, mixed casing, etc.), so I just settled for something generic.
* Tried by couldn't reproduce the duplicate error message problem as we talked about over Slack
* Done button takes you to the correct index pattern on Discover tab

Bargs commented:
Alright @BigFunger I pushed the rest of my updates, I think this is ready for another review. The only thing I haven't addressed is the UI improvement for when you get a mass number of errors on the last page. I'm still waiting for @alt74 to weigh in, I'll ping him on Monday if I don't hear otherwise.

@alt74 for reference, the screenshot belows shows the crazyness you get if things totally blow up on you. Basically what's happening is it's displaying an error for every line in the file. @BigFunger and I talked about a couple different ideas for improving this case like a "See More" link with errors truncated to 5-10 max by default, or possibly pagination. Ideas?

Bargs commented:
Chatted on zoom, 2 items to update:

* Don't pre-fill index name, just add a placeholder
* Update index name help-text so it doesn't imply you can add data to an existing index

BigFunger commented:
@Bargs Has there been any discussion about ignoring certain fields? I am trying to import data from https://public.tableau.com/s/sites/default/files/media/Resources/catsvdogs.xlsx and when I convert it to a CSV in excel, there is an extra column in the result. This ends up with an extra column in the Review step that I can assign a type to, but fails when I try to click save (because the name column is empty) If I could ignore the column, it would fix that issue. You may want to add an additional check for this case because the error that it throws is caught and displayed via notify.

BigFunger commented:
Do we want anything in place to help them convert their data when there are commas in numeric values? This seems like a common scenario.

BigFunger commented:
Changes you applied look good. Please address the issues mentioned in chat (index name, index help text) and the couple other things I mentioned (ignoring fields, numeric conversion)

Bargs commented:
> Do we want anything in place to help them convert their data when there are commas in numeric values? This seems like a common scenario.

@BigFunger did you try mapping those fields as numbers on the pattern review step? I just tried it and Elasticsearch seems to handle the commas fine

Bargs commented:
@BigFunger re: ignoring fields

Would it be acceptable if we just fail early in this scenario? We could detect empty headers on the very first step of the wizard and display an error. Adding a generic "ignore field" functionality would require making the API more complex since the real parsing work is done server side, and I suspect its use cases would overlap some with the remove processor once we've re-added the pipeline step.

Bargs commented:
jenkins, test it

Bargs commented:
jenkins, test it

alt74 commented:
a design update for the error box and truncating would be great. + truncating and then see where this takes us.

Bargs commented:
@BigFunger back to you. The only outstanding issues are the two questions I posed above. Here's what I've updated so far:

* Index name field on the review step is no longer pre-filled.
* Index name field help text no longer implies that you can add data to existing indices.
* Added truncated and toggle-able error messages. Easiest way to test this now is to intentionally map a field incorrectly, e.g. select `number` for a string field which should cause an error for every row.

Note on the error messages: I considered making this a generic directive, but after looking through the code base it seems like the notify service is where this functionality really belongs (or at least a large portion of code should be shared with the notification service). Unfortunately notify would require a pretty big refactoring which would be out of scope for this PR, so rather than confuse the picture even further by adding yet another alerting/notification directive to the code base, I've just implemented a simple solution for this particular page. If we find later that we do need it on other pages, it should be easy to extract.

BigFunger commented:
@Bargs 
> re: ignoring fields - Would it be acceptable if we just fail early in this scenario? We could detect empty headers on the very first step of the wizard and display an error. Adding a generic "ignore field" functionality would require making the API more complex since the real parsing work is done server side, and I suspect its use cases would overlap some with the remove processor once we've re-added the pipeline step.

You're probably right about the ignore field path having overlapping functionality with the Remove processor. I agree that we should detect these field early, but think there are a couple ways we could deal with it:
- Automatically remove it from the input and give the user an info notice
- Fail early and force the user to fix their data and try again
- Assign a default column name to it

Another thing to consider, why do we not allow the user to modify the column names? Again, I anticipate the response of 'this overlaps with functionality from the pipeline', but I'm not sure that we should limit this flow because of that. Importing data via a filebeat and importing data via csv are similar, but also different workflows IMO.

BigFunger commented:
@Bargs 
> did you try mapping those fields as numbers on the pattern review step? I just tried it and Elasticsearch seems to handle the commas fine

Yep. :)

![import-error1](https://cloud.githubusercontent.com/assets/12193008/15553499/fd96ed6c-2284-11e6-98bb-eef67c09cbc0.gif)

(rename to test.csv)
[test.txt](https://github.com/elastic/kibana/files/282803/test.txt)

BigFunger commented:
@Bargs I was able to reproduce the `index by` issue while trying to test the truncated error list.
![temp](https://cloud.githubusercontent.com/assets/12193008/15553735/20ad826a-2286-11e6-8d7a-2f3a16d1e8de.png)

![issue-animation](https://cloud.githubusercontent.com/assets/12193008/15553896/f5c07656-2286-11e6-933d-cc9074cea3ab.gif)

[StormEvents_details-ftp_v1.0_d1977_c20160223.txt](https://github.com/elastic/kibana/files/282828/StormEvents_details-ftp_v1.0_d1977_c20160223.txt)

Bargs commented:
@BigFunger pushed three new updates.

* Fixed the trackby error. Turns out I forgot to update the ID parsing logic in the UI when I changed the ID format that the API generates. Errors should now truly be unique.
* When papaparse reports a parsing error, increment the indicated row by 1 so that it matches the line numbers in the CSV.
* CSV parse step now checks for empty column headers and displays an error message and prevents the user from continuing until they fix their data.

As discussed on Slack, let's hold off on string -> coercion and column renaming/ignoring. We can create separate tickets for these enhancements if we want to keep them in our back pocket.

BigFunger commented on src/plugins/kibana/public/settings/sections/indices/add_data_steps/upload_data_step/upload_data_step.js:
```diff
@@ -28,7 +28,7 @@ modules.get('apps/settings')
             this.created += response.created;
             this.formattedErrors = this.formattedErrors.concat(_.map(_.get(response, 'errors.index'), (doc) => {
-              return `${doc._id.split('-', 1)[0].replace('L', 'Line ').trim()}: ${doc.error.type} - ${doc.error.reason}`;
+              return `Line ${doc._id.substr(doc._id.lastIndexOf(':') + 1)}: ${doc.error.type} - ${doc.error.reason}`;
```
Is it worth pulling this logic into a function? Especially since the format of the data has changed now during the development of this feature?

BigFunger commented:
One more small suggested change, not a deal breaker. Once tests pass, LGTM.

Bargs commented:
jenkins, test it

Bargs commented:
@BigFunger 

* Extracted error formatting into its own function
* Added an API test for ID format

BigFunger commented:
LGTM
![slack-imgs](https://cloud.githubusercontent.com/assets/12193008/15578883/e56c4bc6-2327-11e6-80b3-de985f4301b2.gif)

BigFunger commented:
Still LGTM. Thanks for addressing the duplicate column issue.

w33ble commented:
I have a CSV file with 2,600+ columns and 68,000+ rows. Importing that seems to blow up the parser or something. 

![screenshot 2016-05-27 13 47 09](https://cloud.githubusercontent.com/assets/404731/15621227/52ea00ac-2414-11e6-8831-1010194aabc6.png)

![screenshot 2016-05-27 13 53 23](https://cloud.githubusercontent.com/assets/404731/15621230/577fcfc0-2414-11e6-9825-6b85cc6077af.png)

Also, apparently Elasticsearch doesn't allow that many fields on a document:

![screenshot 2016-05-27 14 03 45](https://cloud.githubusercontent.com/assets/404731/15621224/476f22ac-2414-11e6-9967-09549177274c.png)

Pretty edge-case, I know, just wanted to share.

w33ble commented:
@Bargs since you seemed to think that the large file importing was caused by misuse of papaparse, I'm sending back to you to fix that. 

OTT, LGTM!

Bargs commented:
@w33ble just pushed some performance improvements. Really glad you caught this issue. After profiling the code I realized there were a combination of things dragging performance down. Give it another whirl and tell me what you think. I was able to add your 700MB CSV to the first wizard step in about 8 seconds without the browser choking. As far as I can tell, the bottleneck is Angular now. With all those columns, it's having to create a lot of DOM elements. If you have any suggestions on optimizing that, I'd be all ears!

w33ble commented:
![](https://media.giphy.com/media/it6W8D4FfvaPC/giphy.gif)
